crates/adapters/src/transport/url.rs

```diff
@@ -0,0 +1,248 @@
+        }
+    }
+
+    async fn worker_thread(
+        config: UrlInputConfig,
+        consumer: &mut Box<dyn InputConsumer>,
+        mut receiver: Receiver<PipelineState>,
+    ) -> AnyResult<()> {
+        let client = Client::builder()
+            .connector(Connector::new().rustls(rustls_config()))
+            .finish();
+        let request = client.get(config.path);
+        let mut response = request.send().await.map_err(
+            // `awc` intentionally uses errors that aren't `Sync`, but
+            // `anyhow::Error` requires `Sync`.  Transform the error so we can
+            // return it.
+            |error| anyhow!("{error}"),
+        )?;
+
+        while *receiver.borrow() != PipelineState::Terminated {
+            select! {
+                _ = receiver.changed() => (),
+                result = response.next(), if *receiver.borrow() == PipelineState::Running => {
+                    match result {
+                        None => return Ok(()),
+                        Some(Ok(data)) => consumer.input_fragment(&data)?,
+                        Some(Err(error)) => Err(error)?,
+                    }
+                }
+            }
+        }
+
+        Ok(())
+    }
+}
+
+impl InputEndpoint for UrlInputEndpoint {
+    fn connect(&mut self, mut consumer: Box<dyn InputConsumer>) -> AnyResult<()> {
+        let config = self.config.clone();
+        let receiver = self.receiver.clone();
+        let _worker = spawn(move || {
+            System::new().block_on(async move {
+                let result = Self::worker_thread(config, &mut consumer, receiver).await;
+                if let Some(error) = result.err().or_else(|| consumer.eoi().err()) {
+                    consumer.error(true, error);
+                }
+            });
+        });
+        Ok(())
+    }
+
+    fn pause(&self) -> AnyResult<()> {
+        Ok(self.sender.send(PipelineState::Paused)?)
+    }
+
+    fn start(&self) -> AnyResult<()> {
+        Ok(self.sender.send(PipelineState::Running)?)
+    }
+
+    fn disconnect(&self) {
+        let _ = self.sender.send(PipelineState::Terminated);
+    }
+}
+
+impl Drop for UrlInputEndpoint {
+    fn drop(&mut self) {
+        self.disconnect();
+    }
+}
+
+fn rustls_config() -> Arc<ClientConfig> {
+    lazy_static! {
+        static ref ROOT_STORE: Arc<ClientConfig> = {
+            let mut root_store = RootCertStore::empty();
+            root_store.add_server_trust_anchors(TLS_SERVER_ROOTS.iter().map(|ta| {
+                OwnedTrustAnchor::from_subject_spki_name_constraints(
+                    ta.subject,
+                    ta.spki,
+                    ta.name_constraints,
+                )
+            }));
+
+            Arc::new(
+                ClientConfig::builder()
+                    .with_safe_defaults()
+                    .with_root_certificates(root_store)
+                    .with_no_client_auth(),
+            )
+        };
+    }
+
+    ROOT_STORE.clone()
+}
+
+#[cfg(test)]
```

Comment:
It would be nice to test some corner cases:
- termination/pause/unpause while downloading
- errors: invalid url, server-side disconnect.

Reply:
I added some tests.